reader.go

```diff
@@ -29,31 +66,847 @@ type Reader struct {
+			}
+
+			if err := r.commitOffsets(conn, offsetsByTopicAndPartition); err != nil {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Printf("unable to commit offset: %v", err)
+				})
+				return
+			}
+		}
+	}
+}
+
+func (r *Reader) commitLoopInterval(conn offsetCommitter, stop <-chan struct{}) {
+	ticker := time.NewTicker(r.config.HeartbeatInterval)
+	defer ticker.Stop()
+
+	offsetsByTopicAndPartition := map[string]map[int]int64{}
+
+	defer func() {
+		// commits any outstanding offsets on close
+		r.commitOffsets(conn, offsetsByTopicAndPartition)
+	}()
+
+	for {
+		select {
+		case <-stop:
+			return
+
+		case <-ticker.C:
+			if len(offsetsByTopicAndPartition) == 0 {
+				continue
+			}
+
+			if err := r.commitOffsets(conn, offsetsByTopicAndPartition); err != nil {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Printf("unable to commit offset: %v", err)
+				})
+				return
+			}
+			offsetsByTopicAndPartition = map[string]map[int]int64{}
+
+		case m, ok := <-r.commits:
+			if !ok {
+				r.withErrorLogger(func(l *log.Logger) {
+					l.Println("reader commit channel unexpectedly closed")
+				})
+				return
+			}
+
+			offsetsByPartition, ok := offsetsByTopicAndPartition[m.Topic]
+			if !ok {
+				offsetsByPartition = map[int]int64{}
+				offsetsByTopicAndPartition[m.Topic] = offsetsByPartition
+			}
+
+			if offset, ok := offsetsByPartition[m.Partition]; ok && offset >= m.Offset {
+				// offsets should only increase
+				continue
+			}
+
+			offsetsByPartition[m.Partition] = m.Offset
+		}
+	}
+}
+
+func (r *Reader) commitLoop(stop <-chan struct{}, done chan<- struct{}) {
+	defer close(done)
+
+	r.withLogger(func(l *log.Logger) {
+		l.Println("started commit group,", r.config.GroupID)
+	})
+	defer r.withLogger(func(l *log.Logger) {
+		l.Println("stopped commit group,", r.config.GroupID)
+	})
+
+	conn, err := r.connect(r.stctx)
+	if err != nil {
+		r.withErrorLogger(func(l *log.Logger) {
+			l.Printf("unable to connect to broker for commit loop: %v", err)
+		})
+		return
+	}
+	defer conn.Close()
+
+	if r.config.CommitInterval == 0 {
+		r.commitLoopImmediate(conn, stop)
+	} else {
+		r.commitLoopInterval(conn, stop)
+	}
+}
+
+func (r *Reader) run() {
+	defer close(r.done)
+
+	const (
+		backoffDelayMin = 100 * time.Millisecond
+		backoffDelayMax = 3 * time.Second
+	)
+
+	if r.config.GroupID == "" {
+		return
+	}
+
+	r.withLogger(func(l *log.Logger) {
+		l.Printf("entering loop for consumer group, %v\n", r.config.GroupID)
+	})
+
+	for attempt := 0; true; attempt++ {
+		if attempt != 0 {
+			if !sleep(r.stctx, backoff(attempt, backoffDelayMin, backoffDelayMax)) {
+				return
+			}
+		}
+
+		// rebalance and fetch subscriptions
+		assignments, err := r.rebalance()
+		if err != nil {
+			r.withErrorLogger(func(l *log.Logger) {
+				l.Printf("rebalance failed for consumer group, %v: %v", r.config.GroupID, err)
+			})
+			continue
+		}
+
+		// start the heartbeat
+		hbStop, hbDone := make(chan struct{}), make(chan struct{})
```

Comment:
It seems like a `sync.WaitGroup` would be a better construct to wait for all sub-goroutines to complete instead of waiting on multiple "done" channels.